scripts/stablex/transitive_orderbook.js

```diff
@@ -0,0 +1,100 @@
+const BatchExchangeViewer = artifacts.require("BatchExchangeViewer")
+const { getOpenOrdersPaginated } = require("./utilities.js")
+const BN = require("bn.js")
+
+const { Orderbook, Offer } = require("../../typescript/common/orderbook.js")
+const { Fraction } = require("../../typescript/common/fraction.js")
+
+const argv = require("yargs")
+  .option("sellToken", {
+    describe: "the token you are looking to sell",
+  })
+  .option("buyToken", {
+    describe: "the token you are looking to sell",
+  })
+  .option("sellAmount", {
+    describe: "the amount you are looking to sell",
+  })
+  .option("pageSize", {
+    default: 100,
+    describe: "The page  size for the function getOrdersPaginated",
+  })
+  .demand(["sellToken", "buyToken", "sellAmount"])
+  .version(false).argv
+
+const addItemToOrderbooks = function(orderbooks, item) {
+  let orderbook = new Orderbook(item.sellToken, item.buyToken)
+  if (!orderbooks.has(orderbook.pair())) {
+    orderbooks.set(orderbook.pair(), orderbook)
+  }
+  orderbook = orderbooks.get(orderbook.pair())
+  const price = new Fraction(item.priceNumerator, item.priceDenominator)
+  const volume = new Fraction(item.remainingAmount.gt(item.sellTokenBalance) ? item.sellTokenBalance : item.remainingAmount, 1)
+  // Smaller orders cannot be matched anyways
+  if (volume.gt(new Fraction(10000, 1))) {
+    orderbook.addAsk(new Offer(price, volume))
+  }
+}
+
+const getAllOrderbooks = async function(instance, pageSize) {
+  const elements = await getOpenOrdersPaginated(instance, pageSize)
+  const orderbooks = new Map()
+  elements.forEach(item => {
+    addItemToOrderbooks(orderbooks, item)
+  })
+  return orderbooks
+}
+
+const transitiveOrderbook = function(orderbooks, start, end) {
+  const result = new Orderbook(start, end)
+  // Add the direct book if it exists
+  if (orderbooks.has(result.pair())) {
+    result.add(orderbooks.get(result.pair()))
+  }
+
+  // Check for each orderbook that starts with same baseToken, if there exists a connecting book.
+  // If yes, build transitive closure
+  orderbooks.forEach(book => {
+    const otherBook = orderbooks.get(new Orderbook(book.quoteToken, end).pair())
+    if (book.baseToken === start && otherBook) {
+      const closure = book.transitiveClosure(otherBook)
+      result.add(closure)
+    }
+  })
+  return result
+}
+
+module.exports = async callback => {
+  try {
+    const sellAmount = new BN(argv.sellAmount)
```

Comment:
Most of this logic I think should be moved to a utils function. Imo, only the part of getting the parameters, and handing exception should be done in the script